packages/vite/src/node/plugins/asset.ts

```diff
@@ -251,17 +254,26 @@ export async function fileToUrl(
 ): Promise<string> {
   const { environment } = pluginContext
   if (environment.config.command === 'serve') {
-    return fileToDevUrl(id, environment.getTopLevelConfig())
+    return fileToDevUrl(environment, id)
   } else {
     return fileToBuiltUrl(pluginContext, id)
   }
 }
 
-export function fileToDevUrl(
+export async function fileToDevUrl(
+  environment: Environment,
   id: string,
-  config: ResolvedConfig,
   skipBase = false,
-): string {
+): Promise<string> {
+  const config = environment.getTopLevelConfig()
+
+  // If has inline query, unconditionally inline the asset
+  if (inlineRE.test(id)) {
+    const file = cleanUrl(id)
+    const content = await fsp.readFile(file)
+    return assetToDataURL(environment, file, content)
+  }
+
   let rtn: string
   if (checkPublicFile(id, config)) {
```

Comment:
This seems to be inconsistent with the build behavior.
https://github.com/vitejs/vite/blob/8cda9f97b57e4c30ac9a7d53acde9dd5161bf527/packages/vite/src/node/plugins/asset.ts#L350-L364

I think we don't want `?inline` to be applied to public files. It feels confusing for me.

Reply:
Perhaps build needs to skip the public check if there's the `?inline` query here to match dev 🤔 I'll explain the reasoning for this in your other comment

Reply:
I'd prefer not allowing `?inline` with public files as it makes the difference of public files and non public files blurry, but won't block this PR for that.
If we are going to support it, I think we should handle the `.json`/`.js` + `?url` thing (https://github.com/vitejs/vite/pull/13422, #17915). So the expected behavior is:
```js
import publicJson from '/foo.json' // error happens (already working)
import publicJsonUrl from '/foo.json?url' // an URL is returned (already working)
import publicJsonInline from '/foo.json?inline' // this should error (not sure if it's working)
import publicJsonUrlInline from '/foo.json?inline&url' // a data URL is returned (not sure if it's working)
```

Reply:
I pushed a fix that should make it consistent now with some tests. I think `?inline` for me here should only work for assets, so that includes `.png` or `.json?url`. And both of them can also be imported from public directory without issues too today. So for consistency, supporting `?inline` should also apply to both too. I'm happy to get another opinion on this though.

> ```js
> import publicJsonInline from '/foo.json?inline'
> ```

It now errors as expected as it's not considered an asset without `?url`


> ```js
> import publicJsonUrlInline from '/foo.json?inline&url'
> ```

This should now return a data url

Reply:
Thanks!

Reply:
cc @patak-dev for thoughts on https://github.com/vitejs/vite/pull/15454#discussion_r1825571264

Reply:
Sounds good to me 👍🏼